### ISSUE TITLE
set_script() fails if target node has no parent or if node is root sometimes

### ISSUE THREAD

**Issue by BiRD4** [godotengine/godot#38373]
**Godot version:**
v3.2.1.stable.official

**OS/device including version:**
Windows 10 Home, version 1903, build 18362.778

**Issue description:**
```set_script()``` doesn't work if it's **(1)** called on a node that has been added as a child in a script or if it's **(2)** being called on the root node by a child (created in a script) of the root node. I haven't tested the second part on anything except for the root node, so it might be a broader issue. I've also only tested with a Spatial root node and a Node child node. The first of these problems has a workaround: just add the child after setting the script. The second problem seems to happen as long as the child node was created a script (or at least the root/parent node's initial script), so there's no workaround that I'm aware of.

**Steps to reproduce:**
<details>
<summary><b>(1)</b> Calling set_script() on a Node That Has Been Added as a Child:</summary>
<p>
&emsp;<b>1.</b> Add a Spatial root node and give it a script<br/>
&emsp;<b>2.</b> In the script's _ready(), create a new Node,  add it as a child of the Spatial node, and set the new Node's script to the second script (we will write it in the next step) (make sure everything happens in that order)<br/>
&emsp;<b>3.</b> In the second script's _ready(), print something out to see if it ran
</p>
</details>
<details>
<summary><b>(2)</b> Calling set_script() on the Root Node in the Script of a Node Created by a Script:</summary>
<p>
The same first 2 steps except, in the second step, the new Node is added as a child after the script is set:<br/>
&emsp;<b>1.</b> Add a Spatial root node and give it a script<br/>
&emsp;<b>2.</b> In the script's _ready(), create a new Node, set the new Node's script to the second script (we will write it in the next step), and add it as a child of the Spatial node (make sure everything happens in that order)<br/>
&emsp;<b>3.</b> In the second script's _ready(), print something out to see if it ran, get the parent node (this will return the original Spatial node) and set its script to the third script:<br/>
&emsp;<b>4.</b> In the third script's _ready(), print something out to see if it ran
</p>
</details>

**Minimal reproduction project:**
Each example has its own scene. I've included two working examples to show that the first problem doesn't occur when the new Node is added as a child after the script is set and that the second problem doesn't occur when the new Node is added in the editor as opposed to being added by a script.
[set_script() Issue Examples.zip](https://github.com/godotengine/godot/files/4562384/set_script.Issue.Examples.zip)

**Comment by samdze** [comment]
I don't think `set_script()` fails, it's just that the ready notification has already been called when you change the script of those nodes.
`_ready` is called only one time when a node is added to the scene tree, after `_enter_tree`.
In fact, when you set the script and then add the node to the scene tree, it works as expected.

**Comment by BiRD4** [comment]
I meant to mention this in the OP, but ```_process()``` doesn't work either, at least in my first example.

**Comment by BiRD4** [comment]
>In fact, when you set the script and then add the node to the scene tree, it works as expected.

That's what I did in my minimal reproduction project to demonstrate that the issue occurs only when the node is added as a child before calling ```set_script()```. The problem isn't that ```_ready()``` doesn't run, but rather that ```_process()``` doesn't run either. Sorry for the lack of clarity.

**Comment by samdze** [comment]
Ok if that's the case it should be a bug then.

**Comment by TheDuriel** [comment]
The internal _ready call performs a set_process(true) call IIRC.

**Comment by BiRD4** [comment]
Well, I guess that would explain it. So I suppose there's no bug, but if you want to change a node's script after it's been added to the tree, you should just call ```node.set_process(true)``` where ```node``` is the node whose script is being set?

**Comment by samdze** [comment]
I'd not consider this as expected behaviour however.
I'd expect that a script with a `_process` function defined would run with no further setup/care needed after a `set_script` call.

**Comment by BiRD4** [comment]
Yeah, that's what I was thinking, but I figure there might be a reason to do it this way.

**Comment by ericmeng** [comment]
Hi, I encountered the issue. It seems _ready is some sort of reflection call that iterates through all parent classes. So simply calling _ready doesn't work. 

I found one way is to remove the node, call request_ready() then insert it back. Though I did have to create a new script object, never got it working replacing the existing script source_code and reload. Which means the new script has no memory data of the old script. So I had to convert to data and load it.

Other than that everything seem to work perfectly.

**Comment by MikeSchulze** [comment]
I run also in this issue on my plugin project to spy on a scene.
It dosen't work in general on a Node

Simple step to reproduce;
- create a simple scene with attached script
- load and instanciate the scene
- try to exchange the script, the script is not changed it is now NULL


````
	var resource := load("res://addons/gdUnit3/test/mocker/resources/scenes/TestScene.tscn")
	var instance :Control = resource.instance()
	prints("original script", instance.get_script())
	# load same script as used on original scene
	var new_script = load("res://addons/gdUnit3/test/mocker/resources/scenes/TestScene.gd").new()
	prints("new script", new_script)
	# exchange the script
	instance.set_script(new_script)
	prints("exchanged script", instance.get_script())
````
the output
````
original script [GDScript:1548]
new script [Control:1565]
exchanged script [Object:null]

````

**Comment by solodeveloping** [comment]
Maybe this will be useful (hopefully) to someone, do not confuse _ready and _init, _init is called when calling set_script, not _ready
